core/src/main/scala/kafka/log/LogManager.scala

```
@@ -203,6 +208,11 @@ class LogManager(val logDirs: Array[File],
                          delay = InitialTaskDelayMs,
                          period = flushCheckpointMs,
                          TimeUnit.MILLISECONDS)
+      scheduler.schedule("kafka-delete-logs",
```

I think it is worth observing that an alternate approach to this recurring task that pulls of a queue is to just invoke async delete on demand when we delete a log. i.e., instead of having the `asyncDelete` method and log directory renames, instead call `scheduler.schedule("delete", log.delete)` which will set period -1 to get a single invocation. The reason this isn't very robust is we could get several partitions to delete in a short period of time which could lead to rejected executions if the threadpool is fully utilized.

---

The `deleteLogs` function swallows all exceptions and reports them. On IO exceptions on a particular log, it will retry later. The current implementation has a queue of logs already. It's semantically equivalent to `scheudler.schedule` afaik. I did not feel compelled to change.

---

It is similar, but as noted above I think `scheduler.schedule` could lead to a rejected execution if the executor is saturated; and then you would need a rejected execution handler.

---

The scheduler uses `ScheduledThreadPoolExecutor`, which has a fixed number of threads and an unbounded queue. Using this scheduler for IO risks blocking all the threads in the thread-pool if/when disk IO takes too long or never returns. As this thread-pool executor is used for other periodic tasks such as, log-retention, log-flushing, and recovery-checkpointing, we definitely don't want to lock-up all the threads doing nothing. Direct use of  `scheduler.schedule` opens up this unlikely but dangerous possibility. This is perhaps strong enough reason to stay away from direct scheduling and use an explicit queue of logs such that only one thread from the pool is used at a time.

---

That can be avoided by using a separate scheduler.

---

I might be missing something here, but would like to understand the advantage of scheduling delete on the fly over having it as a periodic task.

---

It's simpler - that is pretty much it.

---

A separate scheduler may need it's own `defaultNumThreads` config?